go/p2p/discovery/peerstore/peerstore.go

```diff
@@ -0,0 +1,502 @@
+	func() {
+		s.mu.RLock()
+		defer s.mu.RUnlock()
+
+		for _, regs := range s.registrations {
+			for _, reg := range regs {
+				if reg.expires.Before(now) {
+					continue
+				}
+				peerMap[reg.addr.ID] = reg.addr
+			}
+		}
+	}()
+
+	peers := make([]peer.AddrInfo, 0, len(peerMap))
+	for _, peer := range peerMap {
+		peers = append(peers, peer)
+	}
+
+	return peers
+}
+
+// Load loads peers from the store.
+func (s *Store) Load() error {
+	// Allow only one save/load at a time.
+	s.storeCh <- struct{}{}
+	defer func() { <-s.storeCh }()
+
+	// Load addresses in json form.
+	bucket, err := s.store.GetServiceStore(storeBucketName)
+	if err != nil {
+		return err
+	}
+	var ps storeBackup
+	if err = bucket.GetCBOR([]byte(storePeerstoreKey), &ps); err != nil {
+		return err
+	}
+
+	for _, nsStore := range ps.Namespaces {
+		for _, json := range nsStore.Addrs {
+			var info peer.AddrInfo
+			err := info.UnmarshalJSON(json)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshal addr info: %w", err)
+			}
+
+			_, err = s.Add(nsStore.Namespace, info)
+			s.logger.Debug("failed to add peer to the store",
+				"err", err,
+			)
+		}
+	}
+
+	return nil
+}
+
+// Save persists peers to the store.
+func (s *Store) Save() error {
+	// Allow only one save/load at a time.
+	s.storeCh <- struct{}{}
+	defer func() { <-s.storeCh }()
+
+	// Gather the data.
+	ps, err := func() (*storeBackup, error) {
+		s.mu.RLock()
+		defer s.mu.RUnlock()
+
+		if len(s.registrations) == 0 {
+			return nil, nil
+		}
+
+		sb := storeBackup{
+			Timestamp:  time.Now(),
+			Namespaces: make([]namespaceBackup, 0, len(s.registrations)),
+		}
+
+		for ns, regs := range s.registrations {
+			nsStore := namespaceBackup{
+				Namespace: ns,
+				Addrs:     make([][]byte, 0, len(regs)),
+			}
+
+			for _, reg := range regs {
+				json, err := reg.addr.MarshalJSON()
+				if err != nil {
+					return nil, err
+				}
+				nsStore.Addrs = append(nsStore.Addrs, json)
+			}
+
+			sb.Namespaces = append(sb.Namespaces, nsStore)
+		}
+
+		return &sb, nil
+	}()
+	if err != nil || ps == nil {
+		return err
+	}
+
+	// Persist addresses.
+	bucket, err := s.store.GetServiceStore(storeBucketName)
+	if err != nil {
+		return err
+	}
+	if err = bucket.PutCBOR([]byte(storePeerstoreKey), *ps); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Start starts background services which periodically backup and clean the store.
+func (s *Store) Start() {
+	s.startOne.TryStart(s.run)
+}
+
+// Stop stops all background services. This method blocks until all services are stopped.
+func (s *Store) Stop() {
+	s.startOne.TryStop()
+}
+
+func (s *Store) run(ctx context.Context) {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		s.startBackups(ctx)
+	}()
+
+	go func() {
+		defer wg.Done()
+		s.startCleanups(ctx)
+	}()
+}
+
+// startBackups periodically backups peers to the common store.
+func (s *Store) startBackups(ctx context.Context) {
```

Comment:
The backup part is very similar to the `peermgmt/backup.go`. Could the common backup functionality be extracted and reused?